uast/uast-encoding-v2.md

```diff
@@ -0,0 +1,249 @@
+
+`last_id` is an optional field that sets a maximal node ID reserved by the node ID allocator.
+If not set, `max(node.ID)` is assumed. Implementation may reserve more IDs than required by
+the UAST nodes by setting `last_id > max(node.ID)`.
+
+`root` is an ID of the tree root node. Multiple unnamed roots can be stored by referencing an
+[Array](#array) node by its ID, and multiple named roots can be stored by referencing an
+[Object](#object) node. If not set explicitly, implementations should search the graph for
+unreferenced nodes of type [Array](#array) or [Object](#object) and treat them as an array of roots.
+
+`metadata` is an optional ID of a file metadata root, similar to `root`. This tree may
+contain arbitrary tree structure that describes the UAST file. If not set, no metadata is
+assumed. Implementation may ignore this tree, except for a case when `root` is not explicitly.
+
+### Nodes list
+
+Node list follows the [Header](#header) message. The list contains zero or more length-delimited
+[Node](https://github.com/bblfsh/sdk/blob/b44cc02214bf0d9c55dc3b830f423a6df6709ca9/uast/nodes/nodesproto/nodes.proto#L46)
+messages until the EOF. There is no explicit node count provided by the format.
+
+### Node
+
+Each Node is described by the following protocol buffer message:
+```proto
+message Node {
+    uint64 id = 1;
+    
+    oneof value {
+        string string = 2;
+        int64  int    = 3;
+        uint64 uint   = 4;
+        double float  = 5;
+        bool   bool   = 6;
+    }
+    
+    repeated uint64 keys = 7;
+    uint64 keys_from = 10;
+    
+    repeated uint64 values = 8;
+    
+    bool is_object = 9;
+    
+    uint64 values_offs = 11;
+}
+```
+
+The `id` field is unique, monotonically increasing value. ID value is used in most message
+fields to reference other nodes.
+
+If `id` field is omitted, the `prevNode.ID + 1` is assumed. The first node with no `id`
+is assumed to have an ID value of `1`.
+
+Nil nodes must not be represented by the Node message. Instead, they are represented by
+ID value `0` when referenced by other nodes.
+
+All other nodes can be of one of 3 kinds: [Values](#values), [Arrays](#array) and [Objects](#object).
+
+#### Values
+
+Values nodes must not have any other fields except `id` (optional) and `value` "oneof" field:
+
+```proto
+message Node {
+    uint64 id = 1;
+    
+    oneof value {
+        string string = 2;
+        int64  int    = 3;
+        uint64 uint   = 4;
+        double float  = 5;
+        bool   bool   = 6;
+    }
+}
+```
+
+Each "oneof" value corresponds to a value type defined in the [UAST v2 representation](./representation-v2.md#values).
+
+Value nodes may be deduplicated when encoding the UAST, thus implementations should make a
+copy of the value when decoding.
+
+#### Array
+
+Array nodes must only have an `id` (optional), `values` and `values_offs` fields set.
+`is_object` field may also be set to `false` when encoding an empty array.
+
+```proto
+message Node {
```

Comment:
(For future consideration): In retrospect, we probably should consider making the mutual exclusion explicit in a later revision:

```proto
message Node {
   uint64 id = 1;
   oneof kind {
      Value value = 2;
      Array array = 3;
      Object object = 4;
   }

   message Value {
      oneof kind {
         string string = 1;
         // …
      }
   }
   message Array {
      uint64 offset = 1;
      repeated uint64 values = 2;
   }
   message Object {
      uint64 offset = 1;
      oneof keys {
         Keys keys = 2;
         uint64 keys_from = 3;
      }
      repeated uint64 values = 4;
   }
   message Keys {
      repeated uint64 keys = 1;
   }
}
```

Even though this duplicates some of the apparent structure, as a design property it's nice to have those type-distinct rather than just usage-distinct. It also makes it harder for consumers to get the field population wrong. (It is, however, a bigger change—and I'm not advocating we block anything for it, just raising the topic since we're paying attention to the code)

Reply:
This was considered and was included in the first implementation, but the protobuf overhead for storing nested fields is too much, assuming the number of nodes in the typical AST. Thus, it was decided to inline fields. This makes decoding a bit harder since it's not handled by Protobuf directly. But since we already have a custom header and the format, it doesn't really matter (in my opinion).

As a note, we were discussing this previously. Maybe worth raising this issue again. But I think it may wait until we have some collaboration with other companies going on.

Reply:
Now that you mention it, I do recall we talked about this back in November.

I believe it will be worthwhile to distinguish between _interchange format_ and _storage format_. Right now we are using the same message type for both. But I'd argue that for an interchange format it is more important to have a reliable API surface (one that is hard to use incorrectly), than to squeeze bits. It would be easy to mechanically transform nested messages (with their extra message-id tags) into flat messages or even something more compact than that (we did this in Kythe, for example https://godoc.org/kythe.io/kythe/go/storage/entryset).

An interchange format has _many_ clients: For us, the drivers produce it, the daemon consumes it, the normalization DSL manipulates it, the protocol clients receive it in responses and convert it to native data structures or JSON. My experience has been that having _only_ textual comments to document which fields have to be set together, and which are mutually exclusive, leads to big maintenance headaches: Updating the message structure to add new features, for example, often requires coordinating atomic changes across clients who all assume a particular set of rules.

For single files, even large ones, the extra bits for size tags almost certainly do not matter for interchange: gRPC compresses protocol messages anyway, and this stuff is all highly compressible. Of course, when we want to store a lot of these, the extra bits _could_ matter (though we should measure it). But then we can define a more compact storage format, private to the archiver, that knows how to go back and forth efficiently between that and interchange.

As you say, though, this is not a blocker for your PR here—only a discussion topic for us to revisit.